File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:
    void dfs(vector<vector<int>>& res, vector<int>& subset, vector<int>& c, int i, int remaining) {
        if (remaining == 0) {
            res.push_back(subset); 
            return;
        } else if (remaining < 0 || i >= c.size()) {
            return;
        }

        subset.push_back(c[i]); 
        dfs(res, subset, c, i, remaining - c[i]); 

        subset.pop_back();
        dfs(res, subset, c, i+1, remaining);  
        
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> s;  
        dfs(res, s, candidates, 0, target); 

        return res; 
    }
};

File: Data Structures & Algorithms/valid-palindrome-ii/submission-0.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        l = 0
        r = len(s)-1
        
        while l < r: 
            if s[l] == s[r]: 
                l += 1
                r -= 1

            else: 
                s1 = s[l+1:r+1]
                s2 = s[l:r]
                if s1 == s1[::-1] or s2 == s2[::-1]:
                    return True
                
                return False
        
        return True